### DISCUSSION TITLE
How to get ScreenViewport by EcInstanceId

### DISCUSSION THREAD

**Discussion by ajay-toobler** [iTwin/itwinjs-core#4273] [Q&A]
I used below code at first to change colour of a element.  which code was running inside `IModelApp.viewManager.onViewOpen.addOnce` so i got `vp` from there.

```ts
 const vp = IModelApp.viewManager.selectedView!;
 const emph = EmphasizeElements.getOrCreate(vp);
      
emph.overrideElements(
  EcInstanceId,
  vp,
  ColorDef.fromString(cloudData[value.smartDeviceId]["Color Status"]),
  FeatureOverrideType.ColorOnly,
  false
);
})

```

after addOnce had run. I need to update again the color of the element having a EcInstanceId. But now how can i get `vp` ?

**Comment by pmconne** [comment]
What is wrong with what you already have (other than failing to handle case when `selectedView` is `undefined`)?

> `const vp = IModelApp.viewManager.selectedView!;`

**Reply by ajay-toobler** [reply]
if it is failing. because of getting undefined.

**Reply by ajay-toobler** [reply]
Is there a way to get `vp` using `EcInstanceId`.

**Reply by pmconne** [reply]
There is no correlation between an element's Id and a viewport.
Can you explain the high-level problem you are trying to solve?

**Reply by ajay-toobler** [reply]
I have seen in this sandbox https://www.itwinjs.org/sandboxes/iTwinPlatform/Emphasize%20Elements . we could change colour of the element by selecting the element using mouse. I thought the background engineering might using EcInstanceId of identifying the element and then selecting it.

**Reply by ajay-toobler** [reply]
I am R&D -ing this package. So i have developed a sample project by looking at https://www.youtube.com/watch?v=BgfbwSNqPlc&list=PL6YCKeNfXXd_dXq4u9vtSFfsP3OTVcL8N&index=1 this playlist.

 my code is similar to https://github.com/iTwin/Course_1-Smart-House-Sample/tree/Chapter_6_Lesson_3 .

Here they used are fetching some smartdevice data from https://smarthomedata.z22.web.core.windows.net/ this endpoint and showing it over the imodal. 

Suppose, what i can do if i want to refetch this api and update the imodal once again without reloading the `<Viewer />` component again.
It is possible that the api data can be placed in a state and updating the state can cause re-rendering the iModal . But that seems so expensive. 

some other queries
- How we can work with streams of data ?
- How we can integrate web-socket ?

**Reply by pmconne** [reply]
> It is possible that the api data can be placed in a state and updating the state can cause re-rendering the iModal . But that seems so expensive.

If by "state" you mean "React state" then you're right - it would be far too expensive to try to let React manage the rendering of the contents of a viewport. The rendering of a viewport's contents has nothing to do with React. Instead, a viewport re-renders whenever something has changed to require it to render. For example, when the mouse cursor moves, all decorators are asked to redraw into the viewport.

The sample to which you linked uses a [Decorator](https://github.com/iTwin/Course_1-Smart-House-Sample/blob/Chapter_6_Lesson_3/src/components/decorators/SmartDeviceDecorator.tsx) to draw a [set of Markers](https://www.itwinjs.org/reference/core-frontend/views/markerset/) representing IoT devices. If you want to re-fetch this API and update the view, you can do that by updating or replacing your decorator's MarkerSet.

> some other queries

How we can work with streams of data ?
How we can integrate web-socket ?

That depends entirely on what you want to do with them. Please feel free to create new discussion topics explaining in detail what you're trying to accomplish.

**Reply by ajay-toobler** [reply]
Can you elaborate this _you can do that by updating or replacing your decorator's MarkerSet._

**Reply by pmconne** [reply]
> Can you elaborate this _you can do that by updating or replacing your decorator's MarkerSet._

[Here's an example](https://www.itwinjs.org/sandboxes/iTwinPlatform/Marker%20Pins) that shows how to add and remove markers.
If you need further assistance please be more specific about what you are trying to do.

**Reply by ajay-toobler** [reply]
Most of the problems i got have solved.

One more to go.

How can I update title of marker inside decorator while hovering the decorator ?

in socket.on inside onIModelConnected
```
if (!mySmartDeviceDecorator) {
          mySmartDeviceDecorator = new SmartDeviceDecorator(
            smartDeviceData,
            data
          );

         IModelApp.viewManager.addDecorator(
            mySmartDeviceDecorator
          );
        } else {
          mySmartDeviceDecorator.updateDecorate(data);
        }
```

inside implemented Decorator class
```
public updateDecorate(cloudData: any) {
    this._markerSet.forEach((marker) => {
      marker.updateTitle(cloudData[marker._smartDeviceId]);
    });
  }
```

inside extended Maker class
```
updateTitle(cloudData:any){
    this.title = this.populateTitle(cloudData);
  }
```

This code works fine. But when I am hovering over the decorator it isn't changing on the UI. I can only see the new data after I hoverOut and hoverOn again. 

How can I lively update decorator title while hovering over the decorator ?

**Reply by pmconne** [reply]
> How can I lively update decorator title while hovering over the decorator ?

@IvanJK97 see [here](https://github.com/iTwin/itwinjs-core/discussions/3512#discussioncomment-3662053).
